hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -3578,15 +3578,23 @@ def __getitem__(self, item):
             raise ValueError(f'Must specify one index per dimension. '
                              f'Expected {self.ndim} dimensions but got {len(item)}')
 
+        def clamp(expr, min, max):
+            return (hl.case()
+                    .when(expr < min, min)
+                    .when(expr > max, max)
+                    .default(expr))
+
         n_sliced_dims = len([s for s in item if isinstance(s, slice)])
         if n_sliced_dims > 0:
             slices = []
             for i, s in enumerate(item):
                 if isinstance(s, slice):
                     step = hl.case().when(s.step != 0, s.step).or_error("Slice step cannot be zero") if s.step is not None else to_expr(1, tint64)
                     start = hl.cond(s.start >= 0, s.start, self.shape[i] + s.start) if s.start is not None else hl.cond(step >= 0, to_expr(0, tint64), self.shape[i] - 1)
+                    clamped_start = clamp(start, 0, self.shape[i] - 1)
                     stop = hl.cond(s.stop >= 0, s.stop, self.shape[i] + s.stop) if s.stop is not None else hl.cond(step >= 0, self.shape[i], to_expr(-1, tint64))
-                    slices.append(hl.tuple((start, stop, step)))
+                    clamped_stop = clamp(stop, -1, self.shape[i])
```

Comment:
should t min here be -self.shape[I]?

Reply:
I don't think so. The long `stop = hl.cond(....` line above should translate everything into positive numbers.

Reply:
I only handle positive starts and stops in the emitter, this cleans them up so we match numpy semantics with indexing from the end using negatives